src/compiler/clang.rs

```diff
@@ -57,49 +58,69 @@ pub fn compile<T>(creator: &T,
                   -> Box<Future<Item=(Cacheable, process::Output), Error=io::Error>>
     where T: CommandCreatorSync,
 {
-    let mut creator = creator.clone();
-    let compiler = compiler.clone();
-    let parsed_args = parsed_args.clone();
-    let cwd = cwd.to_string();
-
-    pool.spawn_fn(move || {
-        trace!("compile");
-        // Clang needs a temporary file for compilation, otherwise debug info
-        // doesn't have a reference to the input file.
+    trace!("compile");
+    // Clang needs a temporary file for compilation, otherwise debug info
+    // doesn't have a reference to the input file.
+    let input = parsed_args.input.clone();
+    let write = pool.spawn_fn(move || {
         let tempdir = try!(TempDir::new("sccache"));
-        let filename = try!(Path::new(&parsed_args.input).file_name().ok_or(Error::new(ErrorKind::Other, "Missing input filename")));
+        let filename = try!(Path::new(&input).file_name().ok_or(Error::new(ErrorKind::Other, "Missing input filename")));
         let input = tempdir.path().join(filename);
-        {
-            try!(File::create(&input)
-                 .and_then(|mut f| f.write_all(&preprocessor_output)))
+        try!(File::create(&input)
+             .and_then(|mut f| f.write_all(&preprocessor_output)));
+        Ok((tempdir, input))
+    });
+
+    let out_file = match parsed_args.outputs.get("obj") {
+        Some(obj) => obj.clone(),
+        None => {
+            return future::err(Error::new(ErrorKind::Other, "Missing object file output")).boxed()
         }
-        let out_file = try!(parsed_args.outputs.get("obj").ok_or(Error::new(ErrorKind::Other, "Missing object file output")));
-        let mut cmd = creator.new_command_sync(&compiler.executable);
+    };
+
+    let compiler2 = compiler.clone();
+    let cwd2 = cwd.to_string();
+    let mut creator2 = creator.clone();
+    let parsed_args2 = parsed_args.clone();
+    let out_file2 = out_file.clone();
```

Comment:
Hm let me see if I can clean this up, I've actually got an idea to make this better.